packages/e2e-tests/specs/editor/various/post-visibility.test.js

```diff
@@ -34,6 +39,41 @@ describe( 'Post visibility', () => {
 
 			expect( currentStatus ).toBe( 'private' );
 		} );
+
+		it( `can be canceled when the viewport is ${ viewport }`, async () => {
+			await setBrowserViewport( viewport );
+
+			await createNewPost();
+
+			await openDocumentSettingsSidebar();
+
+			const initialStatus = await page.evaluate( () => {
+				return wp.data
+					.select( 'core/editor' )
+					.getEditedPostAttribute( 'status' );
+			} );
+
+			await page.click( '.edit-post-post-visibility__toggle' );
+
+			const [ privateLabel ] = await page.$x(
+				'//label[text()="Private"]'
+			);
+			await privateLabel.click();
+			await page.waitForSelector( '.components-confirm-dialog' );
+
+			const cancelButton = await page.waitForSelector(
+				'.components-confirm-dialog button.is-tertiary'
```

Comment:
I'd prefer to use an xpath selection rather than CSS, but for reasons I haven't yet identified,
`page.$x( '//button[text()="Cancel"]' )`
and
`page.waitForXPath( '//button[text()="Cancel"]' )`
both cause Puppeteer's `ElementHandle.click()` to error out stating the node being passed in isn't visible or isn't an HTML element. The only way I've found to get either of those to successfully click the button is to swap out the `click()` method [as described previously](https://github.com/WordPress/gutenberg/pull/37602#issuecomment-1021653467).

So basically it seems like a choice between the alternate `click()` method, or the CSS selector... I'm not sure which would be the lesser of two evils...

Reply:
I'm not an expert of Puppeteer and XPath.

> the node being passed in isn't visible or isn't an HTML element.

This, paired with the previous comment about having to `await`, makes me think that you may be trying to access the Cancel button too early? Maybe the dialog is still opening?

Reply:
I've found out what's wrong: there are actually two buttons on the page with the `Cancel` inner text. We were selecting the wrong one, which happened to not be accessible at the time puppeteer tried to click it (haven't dug into it, though). The fix is as simple as changing the XPath expression to `//div[contains(@class, "components-confirm-dialog")]//button[text()="Cancel"]')`. This still has the downside of relying on a CSS class (`components-confirm-dialog`) which might change in the future but is slightly better than relying on two CSS classes, which is the case with the CSS selector now. We could also traverse to the dialog div by `role` (which has a value of `dialog` now). 

On the other hand, I think it's not a huge deal to leave it with the CSS selector you've used, @chad1008.

Reply:
Instead of using the `components-confirm-dialog` classname, could we use something like `role="dialog"` ?

Alternatively, if/when we will add props for confirm/cancel button labels, we'll be able to add a unique label that will avoid the problem altogether

Reply:
> Instead of using the components-confirm-dialog classname, could we use something like role="dialog" ?
Yep, we can do that 👍🏻

Reply:
> Yep, we can do that 👍🏻

Awesome! Using roles for querying the dom is a great way to add resilience against future changes, and tests the component more from the point of view of the user (instead of testing implementation details)

Reply:
Sorry, wrote too fast :(... I just tested traversing using the `role` attribute and found the same issue. It seems there are more divs on the page with the same role, so it's not specific enough for our use-case :(

Reply:
I think the best trade-off is using the `//div[contains(@class, "components-confirm-dialog")]//button[text()="Cancel"]')` xpath.

Reply:
> It seems there are more divs on the page with the same role, so it's not specific enough for our use-case :(

This may be worth looking into a little bit — it's weird that there are 2 open dialogs in the page (with 2 cancel buttons). It almost sounds as if the `ConfirmDialog` is being rendered twice? And if not the same `ConfirmDialog`, it's still weird that there are 2 open dialogs in the same page?

> I think the best trade-off is to use the `//div[contains(@class, "components-confirm-dialog")]//button[text()="Cancel"]')` xpath.

One last attempt: what about`[role='dialog'] > [role='document'] > button[text='Cancel`]` (of course translated to the correct xpath syntax)?

If nothing else works, we can definitely go with your suggestion. But I'd definitely look a bit more into why there isn't only one "Cancel` button inside a `dialog` in the page.

Reply:
> But I'd definitely look a bit more into why there isn't only one "Cancelbutton inside adialog` in the page.

Found it. It's the core UI link inserter:
![image](https://user-images.githubusercontent.com/13856531/151785401-86a8e914-b2f8-4767-a6b1-8d26536f439b.png)

It gets rendered when the page loads, but is hidden by a couple of `display: none` on `#wp-link-wrap` (one inline style and one added via `editor.css`). The modal (and our phantom button) only actually appear on screen in the Classic Editor when adding a link to text on the page. 

I'll work on an alternate XPath with this in mind and update when I have something!

Reply:
`//*[@role="dialog"][not(@id="wp-link-wrap")]//button[text()="Cancel"]` should do the trick. (https://github.com/WordPress/gutenberg/pull/37602/commits/423ff74056c998361aab7ec73abba24693a4ee4b)

We're relying on the other button's wrapper not to go changing its ID, but I don't know that we have a better way to differentiate between the two. Thoughts? 🙂

Reply:
Of all alternatives, this is my favourite! This test will break only if:

- the dialog's id changes
- a new dialog with a "Cancel" button is added

In both scenarios, it should be quite easy to understand that the change caused the test failure.

I also think that we should update this test as soon as we introduce props for changing the confirm/cancel button labels on `ConfirmDialog`:

> Alternatively, if/when we will add props for confirm/cancel button labels, we'll be able to add a unique label that will avoid the problem altogether